---
title: "Mitochondrial companion- Upset "
output:
  html_document: default
  pdf_document: default
date: "`r Sys.Date()`"
author: "Archana Raja"
---

# Mitochondria companion upset plots (Transcriptomics and Proteomics)

## Set up

```{r include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE)
#Setup------------------------------------------------
library(rjson)
library(MotrpacRatTraining6mo)
library(MotrpacRatTraining6moData)
library(plyr)
library(tidyverse)
library(MotrpacBicQC)
library(imputeTS)
library(ggplot2)
library(ComplexUpset)
library(plyr)

```

Read configuration file

```{r include=FALSE,message=FALSE,warning=FALSE}
# Create a copy of config.yaml in your local directory with your own values 
# and add the path to the config file here
config <- rjson::fromJSON(file = "~/work/configs/mitochondria_pass1b_network_analysis-config.json")
date <- Sys.Date()
date <- gsub("-", "", date)

```

Loading zone

```{r include=FALSE,message=FALSE,warning=FALSE}
###  Set working directory ### 
setwd(file.path(config$setwd_user, "mitochondria"))

# path to local clone of https://github.com/MoTrPAC/motrpac-mawg
# gitdir_mawg = "~/GitHub/motrpac-mawg"
# gitdir_mawg <- config$gitdir_mawg

# Load graphical cluster analysis
# 'gsutil' is the path to your gsutil binary
# gsutil may have been installed somewhere else on your system
# Run 'which gsutil' from the Terminal and copy/paste the path into this variable
# gsutil = "/usr/bin/gsutil"
# gsutil <- path.expand(config$gsutil_user)

# 'outdir_scratch' is an intermediate folder where files are downloaded from GCP locally
# before loading them into R
outdir_scratch <- path.expand(config$outdir_scratch)

## source functions
# source(sprintf("%s/pass1b-06/integrative/clustering/cluster_viz_fx.R", gitdir_mawg))
# source(sprintf("%s/pass1b-06/integrative/clustering/graphical_analysis_functions.R", gitdir_mawg))

```

```{r}
#load and subset mitochondrial genes
#to do move the below path to config
mito_genes = read.delim("/Users/araja7/work/companion/mitochondria/scratch/input/mito-genes_gencode_merged_v1.txt",sep="\t",header=TRUE,fill=TRUE)
mito_genes <- mito_genes %>%
  select("entrez_gene","rgd_gene","gene_symbol","ensembl_gene","old_gene_symbol") %>% as.data.frame()
#below yielded 1346 unique genes
mito_genes <- mito_genes %>% distinct() 
#subset the training regulated features to TRNSCRPT 
rnaseq_training <- subset(TRAINING_REGULATED_FEATURES, assay %in% c('TRNSCRPT'))

#subset to mitogenes
dea_merged = merge(x = rnaseq_training, y = mito_genes, by.x ="feature_ID",by.y="ensembl_gene")
#slice df
rnaseq_training_sub <- dea_merged[ , c('feature_ID', 'tissue')]
df_sub=distinct(rnaseq_training_sub)
write.table(df_sub,"/Users/araja7/work/companion/mitochondria/scratch/output/test_df_sub.tsv",row.names=FALSE,quote=F,sep="\t")

```


```{r}
#preparing data for upset (rna-seq)

df_final=df_sub

for (i in unique(df_sub$tissue)) {
    df_slice=df_sub[df_sub$tissue == i,]
    print(head(df_slice))
    df_slice=df_slice[,1, drop=FALSE]
    final_df_slice <- df_slice %>% group_by(.data[["feature_ID"]]) %>%
  summarise(n = n()) %>% as.data.frame()
    colnames(final_df_slice)[2] <- i
    df_final = merge(x = df_final, y = final_df_slice, by ="feature_ID",all.x=TRUE)
}
#remove the tissue column
df_final=(subset(df_final, select = -c(tissue)))
#remove duplicate rows with identical feature_ids
df_final=distinct(df_final)
write.table(df_final,"/Users/araja7/work/companion/mitochondria/scratch/output/test_df_final_for_upset.tsv",row.names=FALSE,quote=F,sep="\t")

#replace all NAs with zero
df_final=na_replace(df_final, 0)
#replace >1 values with 1
df_final <- df_final %>% mutate_if(is.numeric, ~1 * (. != 0))
#df_final_with_annot <- merge(x = df_final, y = mito_genes, by.x ="feature_ID",by.y = "ensembl_gene",all.x=TRUE)
#write.table(df_final_with_annot,"/Users/araja7/work/companion/mitochondria/scratch/output/significant_genes_training.txt",row.names=FALSE,quote=F,sep="\t")

#upset plot
set_names<-names(df_final)
set_names <- set_names[! set_names %in% c('feature_ID')]
#subset to specific set names
set_names=c("ADRNL","BAT","COLON","SKM-GN","SKM-VL","LIVER","HEART","WAT-SC","BLOOD")
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_rnaseq_subset_min6.pdf",width=12,height=5)

upset(
    df_final,set_names,width_ratio=0.2, min_size=6,
#    sort_intersections_by=c('cardinality','degree'),
    base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.5))),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='ADRNL', fill=tissue_cols[["ADRNL"]],only_components=c('intersections_matrix')),
        upset_query(set='BAT', fill=tissue_cols[["BAT"]],only_components=c('intersections_matrix')),
        upset_query(set='COLON', fill=tissue_cols[["COLON"]],only_components=c('intersections_matrix')),
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='LIVER', fill=tissue_cols[["LIVER"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"],only_components=c('intersections_matrix')),
        upset_query(set='WAT-SC', fill=tissue_cols[["WAT-SC"]],only_components=c('intersections_matrix')),
        upset_query(set='BLOOD', fill=tissue_cols[["BLOOD"]],only_components=c('intersections_matrix'))
        
    )
    
        
    
)


dev.off()

```
```{r}
#upset plot for the 3 muscle tissues
set_names=c("SKM-GN","SKM-VL","HEART")
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_rnaseq_muscle_tissues.pdf",width=12,height=5)

upset(
    df_final,set_names,width_ratio=0.2, min_size=1,min_degree=1,
    #mode = "exclusive_intersection",
    #base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.2))),
    sort_intersections_by=c('cardinality','degree'),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"],only_components=c('intersections_matrix'))
        
    )
    
        
    
)


dev.off()

```
```{r}
#Upset for publication
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_rnaseq_subset.pdf",width=12,height=5)
upset(
    df_final,set_names,width_ratio=0.2, min_size=2,
    queries=list(
        upset_query(
            intersect=c('SKM-VL','COLON','BAT','ADRNL'),
            color='blue',
            fill='blue',
            only_components=c('intersections_matrix', 'Intersection size')
        ),
        upset_query(
            intersect=c('SKM-VL','COLON','BAT','ADRNL','SKM-GN','BLOOD'),
            color='pink',
            fill='pink',
            only_components=c('intersections_matrix', 'Intersection size')
        ),
        upset_query(
            intersect=c('HEART','BAT','ADRNL'),
            color='yellow',
            fill='yellow',
            only_components=c('intersections_matrix', 'Intersection size')
        ),
        upset_query(
            intersect=c('COLON','BAT','ADRNL'),
            color='purple',
            fill='purple',
            only_components=c('intersections_matrix', 'Intersection size')
        ),
                upset_query(
            intersect=c('ADRNL'),
            color='#DDA0DD',
            fill='#DDA0DD',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('SKM-VL'),
            color='#025939',
            fill='#025939',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('BAT'),
            color='#8c5220',
            fill='#8c5220',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('LIVER'),
            color='#da6c75',
            fill='#da6c75',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('COLON'),
            color='#BECA55',
            fill='#BECA55',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('WAT-SC'),
            color='#214da6',
            fill='#214da6',
            only_components=c('intersections_matrix')
        ),
        upset_query(
            intersect=c('BLOOD'),
            color='#d92c04',
            fill='#d92c04',
            only_components=c('intersections_matrix')
        )
        
    ),
                      base_annotations=list(
                      'Intersection size'=(
                        intersection_size(
                          bar_number_threshold=1,  # show all numbers on top of bars
                          width=0.5,   # reduce width of the bars
                        )
                        # add some space on the top of the bars
                        + scale_y_continuous(expand=expansion(mult=c(0, 0.05)))
                        + theme(
                          # hide grid lines
                          panel.grid.major=element_blank(),
                          panel.grid.minor=element_blank(),
                          # show axis lines
                          axis.line=element_line(colour='black')
                        )
                      )
                    )
)


dev.off()
```

```{r}
#qc to check the number of unique and shared genes by counting the number of non-zero rows for each tissue by feature_ids

nonzero <- function(x) sum(x != 0)
numcolwise(nonzero)(df_final)

```
```{r}
#testing
library(ComplexUpset)
set_names=c("SKM-GN","SKM-VL","HEART")
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_rnaseq_subset_test.pdf",width=12,height=5)

upset(
    df_final, 
    set_names,
    width_ratio=0.2,
    group_by='sets',
    min_degree=1,
    max_degree=3,
    queries=list(
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"],only_components=c('intersections_matrix'))
        

    )
)
dev.off()
```
##test2
set_names=c("SKM-GN","SKM-VL","HEART")
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_rnaseq_subset_test.pdf",width=12,height=5)

upset(
    df_final, 
    set_names,
    width_ratio=0.2,
#    group_by='sets',
    min_degree=1,
    max_degree=6,
#    mode='inclusive_union',
    min_size=7,
    queries=list(
        upset_query(set='ADRNL', fill=tissue_cols[["ADRNL"]]),
#        upset_query(set='BAT', fill=tissue_cols[["BAT"]]),
#        upset_query(set='COLON', fill=tissue_cols[["COLON"]]),
        upset_query(set='HEART', fill=tissue_cols[["HEART"]]),
#        upset_query(set='LIVER', fill=tissue_cols[["LIVER"]]),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]]),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"])
#        upset_query(set='WAT-SC', fill=tissue_cols[["WAT-SC"]]),
#        upset_query(set='BLOOD', fill=tissue_cols[["BLOOD"]])
    )
)
dev.off()

```
```{r}
#proteomics
#subset the training regulated features to PROT 
prot_training <- subset(TRAINING_REGULATED_FEATURES, assay %in% c('PROT'))
prot_dea_with_feat = merge(x = prot_training, y = FEATURE_TO_GENE, by.x ="feature_ID",by.y="feature_ID")
#subset to mito genes
prot_dea_with_mito = merge(x = prot_dea_with_feat, y = mito_genes, by.x ="entrez_gene",by.y="entrez_gene")

#slice df
prot_training_sub <- prot_dea_with_mito[ , c('feature_ID', 'tissue')]
df_prot_sub=distinct(prot_training_sub)
write.table(df_prot_sub,"/Users/araja7/work/companion/mitochondria/scratch/output/test_df_prot_sub.tsv",row.names=FALSE,quote=F,sep="\t")

#preparing data for upset (rna-seq)

df_final=df_prot_sub

for (i in unique(df_prot_sub$tissue)) {
    df_slice=df_prot_sub[df_prot_sub$tissue == i,]
    print(head(df_slice))
    df_slice=df_slice[,1, drop=FALSE]
    final_df_slice <- df_slice %>% group_by(.data[["feature_ID"]]) %>%
  summarise(n = n()) %>% as.data.frame()
    colnames(final_df_slice)[2] <- i
    df_final = merge(x = df_final, y = final_df_slice, by ="feature_ID",all.x=TRUE)
}
#remove the tissue column
df_final=(subset(df_final, select = -c(tissue)))
#remove duplicate rows with identical feature_ids
df_final=distinct(df_final)
write.table(df_final,"/Users/araja7/work/companion/mitochondria/scratch/output/test_df_final_for_upset_prot.tsv",row.names=FALSE,quote=F,sep="\t")

#replace all NAs with zero
df_final=na_replace(df_final, 0)
#replace >1 values with 1
df_final <- df_final %>% mutate_if(is.numeric, ~1 * (. != 0))

#upset plot
set_names<-names(df_final)
set_names <- set_names[! set_names %in% c('feature_ID')]
#subset to specific set names
set_names=c("SKM-GN","LIVER","HEART","WAT-SC")
pdf(file="/Users/araja7/work/companion/mitochondria/scratch/output/upset_significant_mito_genes_training_prot_subset_min3.pdf",width=12,height=5)

upset(
    df_final,set_names,width_ratio=0.2, min_size=3,
#    sort_intersections_by=c('cardinality','degree'),
    base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.5))),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='LIVER', fill=tissue_cols[["LIVER"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='WAT-SC', fill=tissue_cols[["WAT-SC"]],only_components=c('intersections_matrix'))
 #       upset_query(set='KIDNEY', fill=tissue_cols[["KIDNEY"]],only_components=c('intersections_matrix'))
  #      upset_query(set='LUNG', fill=tissue_cols[["LUNG"]],only_components=c('intersections_matrix')),
  #      upset_query(set='CORTEX', fill=tissue_cols[["CORTEX"]],only_components=c('intersections_matrix'))
        
        
    )
    
        
    
)


dev.off()

```
```{r}
#qc to make sure the numbers match with those on the heatmap
nonzero <- function(x) sum(x != 0)
numcolwise(nonzero)(df_final)
#  SKM-GN HEART WAT-SC LUNG LIVER KIDNEY CORTEX
#   296   191     94   20   258     50      4
```
