---
title: "Mitochondrial companion- Upset "
output:
  html_document: default
  pdf_document: default
date: "`r Sys.Date()`"
author: "Archana Raja"
---

# Mitochondria companion upset plots (Transcriptomics and Proteomics)

## Set up

```{r include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(message=FALSE,warning=FALSE)
#Setup------------------------------------------------
library(rjson)
library(MotrpacRatTraining6mo)
library(MotrpacRatTraining6moData)
library(plyr)
library(tidyverse)
library(MotrpacBicQC)
library(imputeTS)
library(ggplot2)
library(ComplexUpset)
library(plyr)

```

Read configuration file

```{r include=FALSE,message=FALSE,warning=FALSE}
# Create a copy of config.yaml in your local directory with your own values 
# and add the path to the config file here
config <- rjson::fromJSON(file = "~/work/configs/mitochondria_pass1b_network_analysis-config.json")
date <- Sys.Date()
date <- gsub("-", "", date)

```

Loading zone

```{r include=FALSE,message=FALSE,warning=FALSE}
###  Set working directory ### 
setwd(file.path(config$setwd_user))
#path to github dir
github_dir <- path.expand(config$setwd_user)
#path that contains input files
indir_scratch <- path.expand(config$indir_scratch)
# 'outdir_scratch' is an intermediate folder where files are downloaded from GCP locally
# before loading them into R
outdir_scratch <- path.expand(config$outdir_scratch)

```

```{r}
#load and subset mitochondrial genes
mito_genes_file=file.path(paste0("data/mito-genes_gencode_merged_v1.txt"))
mito_genes = read.delim(mito_genes_file,sep="\t",header=TRUE,fill=TRUE)
mito_genes <- mito_genes %>%
  select("entrez_gene","rgd_gene","gene_symbol","ensembl_gene","old_gene_symbol") %>% as.data.frame()
#below yielded 1346 unique genes
mito_genes <- mito_genes %>% distinct() 
#subset the training regulated features to TRNSCRPT 
rnaseq_training <- subset(TRAINING_REGULATED_FEATURES, assay %in% c('TRNSCRPT'))

#subset to mitogenes
dea_merged = merge(x = rnaseq_training, y = mito_genes, by.x ="feature_ID",by.y="ensembl_gene")
#slice df
rnaseq_training_sub <- dea_merged[ , c('feature_ID', 'tissue')]
df_sub=distinct(rnaseq_training_sub)
###testing only
#write.table(df_sub,file.path(config$outdir_scratch,paste0("test_df_sub.tsv")),row.names=FALSE,quote=F,sep="\t")

```


```{r}
#preparing data for upset (rna-seq)

df_final=df_sub

for (i in unique(df_sub$tissue)) {
    df_slice=df_sub[df_sub$tissue == i,]
    print(head(df_slice))
    df_slice=df_slice[,1, drop=FALSE]
    final_df_slice <- df_slice %>% group_by(.data[["feature_ID"]]) %>%
  summarise(n = n()) %>% as.data.frame()
    colnames(final_df_slice)[2] <- i
    df_final = merge(x = df_final, y = final_df_slice, by ="feature_ID",all.x=TRUE)
}
#remove the tissue column
df_final=(subset(df_final, select = -c(tissue)))
#remove duplicate rows with identical feature_ids
df_final=distinct(df_final)
write.table(df_final,file.path(config$outdir_scratch,paste0("rnaseq_df_final_for_upset.tsv")),row.names=FALSE,quote=F,sep="\t")

#replace all NAs with zero
df_final=na_replace(df_final, 0)
#replace >1 values with 1
df_final <- df_final %>% mutate_if(is.numeric, ~1 * (. != 0))
df_final_with_annot <- merge(x = df_final, y = mito_genes, by.x ="feature_ID",by.y = "ensembl_gene",all.x=TRUE)
write.table(df_final_with_annot,file.path(config$outdir_scratch,paste0("significant_genes_training_with_annot.txt")),row.names=FALSE,quote=F,sep="\t")

#upset plot
set_names<-names(df_final)
set_names <- set_names[! set_names %in% c('feature_ID')]
#subset to specific set names
set_names=c("ADRNL","BAT","COLON","SKM-GN","SKM-VL","LIVER","HEART","WAT-SC","BLOOD")

# Remove rows with zeros in both column A and B
df_filtered <- df_final[!(df_final$`SKM-GN` == 0 & df_final$`LIVER` == 0 & df_final$`HEART` == 0 & df_final$`WAT-SC` == 0 & df_final$`ADRNL` == 0 & df_final$`BAT` == 0 & df_final$`COLON` == 0 & df_final$`SKM-VL` == 0 & df_final$`BLOOD` ==0 ), ]

pdf(file=file.path(config$outdir_scratch,paste0("upset_significant_mito_genes_training_rnaseq_subset_min6_cleaned.pdf")),width=12,height=6)

upset(
    df_filtered,set_names,width_ratio=0.2, min_size=6,
#    sort_intersections_by=c('cardinality','degree'),
    base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.5))),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='ADRNL', fill=tissue_cols[["ADRNL"]],only_components=c('intersections_matrix')),
        upset_query(set='BAT', fill=tissue_cols[["BAT"]],only_components=c('intersections_matrix')),
        upset_query(set='COLON', fill=tissue_cols[["COLON"]],only_components=c('intersections_matrix')),
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='LIVER', fill=tissue_cols[["LIVER"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"],only_components=c('intersections_matrix')),
        upset_query(set='WAT-SC', fill=tissue_cols[["WAT-SC"]],only_components=c('intersections_matrix')),
        upset_query(set='BLOOD', fill=tissue_cols[["BLOOD"]],only_components=c('intersections_matrix'))
        
    )
    
        
    
)


dev.off()

```
```{r}
#upset plot for the 3 muscle tissues
set_names=c("SKM-GN","SKM-VL","HEART")
pdf(file=file.path(config$outdir_scratch,paste0("upset_significant_mito_genes_training_rnaseq_muscle_tissues.pdf")),width=12,height=5)

upset(
    df_final,set_names,width_ratio=0.2, min_size=1,min_degree=1,
    #mode = "exclusive_intersection",
    #base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.2))),
    sort_intersections_by=c('cardinality','degree'),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-VL', fill=tissue_cols["SKM-VL"],only_components=c('intersections_matrix'))
        
    )
    
        
    
)


dev.off()

```

```{r}
#qc to check the number of unique and shared genes by counting the number of non-zero rows for each tissue by feature_ids

nonzero <- function(x) sum(x != 0)
numcolwise(nonzero)(df_final)

```

```{r}
#proteomics
#subset the training regulated features to PROT 
prot_training <- subset(TRAINING_REGULATED_FEATURES, assay %in% c('PROT'))
prot_dea_with_feat = merge(x = prot_training, y = FEATURE_TO_GENE, by.x ="feature_ID",by.y="feature_ID")

#subset to mito genes
prot_dea_with_mito = merge(x = prot_dea_with_feat, y = mito_genes, by.x ="entrez_gene",by.y="entrez_gene")

#slice df
prot_training_sub <- prot_dea_with_mito[ , c('feature_ID', 'tissue')]
df_prot_sub=distinct(prot_training_sub)
###testing 
#write.table(df_prot_sub,file.path(config$outdir_scratch,paste0("test_df_prot_sub.tsv")),row.names=FALSE,quote=F,sep="\t")

#preparing data for upset (proteomics)

df_final=df_prot_sub

for (i in unique(df_prot_sub$tissue)) {
    df_slice=df_prot_sub[df_prot_sub$tissue == i,]
    print(head(df_slice))
    df_slice=df_slice[,1, drop=FALSE]
    final_df_slice <- df_slice %>% group_by(.data[["feature_ID"]]) %>%
  summarise(n = n()) %>% as.data.frame()
    colnames(final_df_slice)[2] <- i
    df_final = merge(x = df_final, y = final_df_slice, by ="feature_ID",all.x=TRUE)
}
#remove the tissue column
df_final=(subset(df_final, select = -c(tissue)))
#remove duplicate rows with identical feature_ids
df_final=distinct(df_final)
write.table(df_final,file.path(config$outdir_scratch,paste0("df_final_for_upset_prot.tsv")),row.names=FALSE,quote=F,sep="\t")

#replace all NAs with zero
df_final=na_replace(df_final, 0)
#replace >1 values with 1
df_final <- df_final %>% mutate_if(is.numeric, ~1 * (. != 0))
#to get the mito genes the df needs to be merged with the master feature to gene map
prot_dea_with_feat = merge(x = df_final, y = FEATURE_TO_GENE, by.x ="feature_ID",by.y="feature_ID")
df_final_with_annot <- merge(x = prot_dea_with_feat, y = mito_genes, by.x ="entrez_gene",by.y = "entrez_gene",all.x=TRUE)
write.table(df_final_with_annot,file.path(config$outdir_scratch,paste0("significant_genes_training_with_annot_prot.txt")),row.names=FALSE,quote=F,sep="\t")
#upset plot
set_names<-names(df_final)
set_names <- set_names[! set_names %in% c('feature_ID')]
#subset to specific set names
set_names=c("SKM-GN","LIVER","HEART","WAT-SC")

# Remove rows with zeros in both column A and B
df_filtered <- df_final[!(df_final$`SKM-GN` == 0 & df_final$`LIVER` == 0 & df_final$`HEART` == 0 & df_final$`WAT-SC` == 0 ), ]

# Print the resulting data frame
print(dim(df_filtered))

pdf(file=file.path(outdir_scratch,paste0("upset_significant_mito_genes_training_prot_subset_min3_test.pdf")),width=12,height=6)
#generate upset displaying sets with minimum 3 genes
upset(
    df_filtered,set_names,width_ratio = 0.2, min_size=3,keep_empty_groups=FALSE,
    base_annotations=list('Size'=(intersection_size(bar_number_threshold = 1,width=0.2))),
    matrix=(
        intersection_matrix(geom=geom_point(shape='circle filled', size=3))
        + scale_color_manual(
            values= tissue_cols,
            guide=guide_legend(override.aes=list(shape='circle'))
        ) 
    ),
    queries=list(
        upset_query(set='HEART', fill=tissue_cols[["HEART"]],only_components=c('intersections_matrix')),
        upset_query(set='LIVER', fill=tissue_cols[["LIVER"]],only_components=c('intersections_matrix')),
        upset_query(set='SKM-GN', fill=tissue_cols[["SKM-GN"]],only_components=c('intersections_matrix')),
        upset_query(set='WAT-SC', fill=tissue_cols[["WAT-SC"]],only_components=c('intersections_matrix'))
        
    )
    
        
    
)


dev.off()

```
```{r}
#qc to make sure the numbers match with those on the heatmap
nonzero <- function(x) sum(x != 0)
numcolwise(nonzero)(df_final)
#  SKM-GN HEART WAT-SC LUNG LIVER KIDNEY CORTEX
#   296   191     94   20   258     50      4
```
