---
title: "Comparison to external datasets"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
---

```{r}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(MotrpacRatTraining6moData)
library(gprofiler2)
disease_datasets_dir = "~/Desktop/repos/motrpac-rat-training-mitochondria/disease_datasets/"
data_dir = "~/Desktop/repos/motrpac-rat-training-mitochondria/data/"
```

Some useful functions for the analyses below

```{r}
simple_enrichment_analysis<-function(s,univ,pathways,min_size=5,setname=""){
  s = intersect(s,univ)
  pathways = lapply(pathways,intersect,y=univ)
  if(length(s)<min_size){
    print("Warning: input set is smaller than min_size, returning NULL")
    return(NULL)
  }
  v1 = univ %in% s
  enrichment_results = c()
  for(pname in names(pathways)){
    pg = pathways[[pname]]
    pg = intersect(pg,univ)
    if(length(pg)<min_size){next}
    v2 = univ %in% pg
    fisher_res = fisher.test(table(v1,v2),alternative = "g")
    overlap_set = intersect(s,pg)
    res = data.frame(
        setname = setname,
        pathway = pname,
        setsize = length(s),pathwaysize=length(pg),
        overlap = length(overlap_set),
        p = fisher_res$p.value,
        genes = paste(overlap_set,collapse=",")
    )
    enrichment_results = rbind(enrichment_results,res)
  }
  return(enrichment_results)
}

add_overlap_tests_and_enrichments<-function(
  dataset_d,motrpac_d,
  dataset_d_gene_set = NULL, # if we have the set of tested genes in another supp table
  motrpac_gene_col="HUMAN_ORTHOLOG_SYMBOL",
  dataset_qval_col = "q",
  dataset_fc_col = "fc",
  dataset_gene_col = "symbol",
  datasetname,tissue,
  min_size_for_sign_analysis = 20,
  fdr_motrpac=0.2,fdr_external=0.2,
  enrichment_species = "hsapiens"){
  
  # Compare fold change sign among significant genes
  set2_male = motrpac_d[motrpac_d$comparison_group == "8w" & motrpac_d$sex == "male",]
  set2_male = set2_male[set2_male$selection_fdr < fdr_motrpac,]
  set2_male = tapply(set2_male$logFC,set2_male[[motrpac_gene_col]],
      FUN = function(x)x[abs(x)==max(abs(x))][1])
  set2_female = motrpac_d[motrpac_d$comparison_group == "8w" & motrpac_d$sex == "female",]
  set2_female = set2_female[set2_female$selection_fdr < fdr_motrpac,]
  set2_female = tapply(set2_female$logFC,set2_female[[motrpac_gene_col]],
      FUN = function(x)x[abs(x)==max(abs(x))][1])
  set2_male = set2_male[names(set2_female)]
  # plot(set2_male,set2_female,pch=20,col="green",cex=0.5,
  #      main="SKMGN PROT: male vs. female fold change",
  #      xlab="Male log2FC",ylab = "Female log2FC");abline(0,1,lty=2,lwd=2)
  
  set1 = dataset_d[dataset_d[[dataset_qval_col]] < fdr_external,]
  # FC1: the fold changes of the selected genes from the external database
  fc1 = tapply(set1[[dataset_fc_col]],set1[[dataset_gene_col]],
               function(x)x[abs(x)==max(abs(x))][1])
  
  # Test for overall overlap among the selected gene sets from the two studies
  all_shared_genes = intersect(unique(dataset_d[[dataset_gene_col]]),
                               unique(motrpac_d[[motrpac_gene_col]]))
  if(!is.null(dataset_d_gene_set)){
      all_shared_genes = union(all_shared_genes,
        intersect(dataset_d_gene_set,unique(motrpac_d[[motrpac_gene_col]])))
  }

  selection1 = all_shared_genes %in% set1[[dataset_gene_col]]
  selection2 = all_shared_genes %in% motrpac_d[!is.na(motrpac_d$selection_fdr) & 
      motrpac_d$selection_fdr<fdr_motrpac,motrpac_gene_col]
  tb = table(selection1,selection2)
  if(!is.null(dim(tb)) && length(c(tb))==4){
    print("Adding overlap test results")
    overall_test = fisher.test(tb,alternative = "g")$p.value
    dataset_overlap_stats = rbind(dataset_overlap_stats,
        data.frame(dataset=datasetname,test = "selection_overlap_p",
        bg_size = sum(tb),overlap=tb["TRUE","TRUE"],
        p=overall_test,tissue=tissue))
    assign("dataset_overlap_stats", dataset_overlap_stats, envir = .GlobalEnv)
  }else{
    print("Skipping overlap test, dataset may have selected features only")
  }
  
  # FC2: male or female MoTrPAC fold changes
  fc2 = set2_male
  fc2 = fc2[!is.na(fc2)]
  shared_genes = all_shared_genes[selection1 & selection2]
  
  if(length(shared_genes) < min_size_for_sign_analysis){
    print("Not enough overlapping selected genes, skipping the sign tests")
    return(NULL)
  }
  
  sign_table = table(
    fc1[shared_genes] > 0,fc2[shared_genes] > 0
  )
  if(length(sign_table)>2){
    dataset_direction_stats = rbind(dataset_direction_stats,
      data.frame(dataset=datasetname,test = "sign_test_male",tissue=tissue,
      bg_size = sum(sign_table),overlap=sign_table["TRUE","TRUE"],
      same_direction = sign_table["TRUE","TRUE"] + sign_table["FALSE","FALSE"],
      opposite_direction = sign_table["TRUE","FALSE"] + sign_table["FALSE","TRUE"],
      p=fisher.test(sign_table,alt="l")$p.value))
  }else{
    dataset_direction_stats = rbind(dataset_direction_stats,
      data.frame(dataset=datasetname,test = "sign_test_male",tissue=tissue,
      bg_size = sum(sign_table),overlap=0,
      same_direction = 0,
      opposite_direction = 0,
      p=NA))
  }
  
  bg = shared_genes
  selected_sets = list(
    male_up_disease_down = shared_genes[
    fc1[shared_genes] < 0 & fc2[shared_genes] > 0],
    male_down_disease_up = shared_genes[
    fc1[shared_genes] > 0 & fc2[shared_genes] < 0],
    male_down_disease_down = shared_genes[
    fc1[shared_genes] < 0 & fc2[shared_genes] < 0],
    male_up_disease_up = shared_genes[
    fc1[shared_genes] > 0 & fc2[shared_genes] > 0]
  )
  
  # Same for females
  fc2 = set2_female
  fc2 = fc2[!is.na(fc2)]
  shared_genes = all_shared_genes[selection1 & selection2]
  sign_table = table(
    fc1[shared_genes] > 0,fc2[shared_genes] > 0
  )
  if(length(sign_table)>2){
    dataset_direction_stats = rbind(dataset_direction_stats,
      data.frame(dataset=datasetname,test = "sign_test_female",tissue=tissue,
      bg_size = sum(sign_table),overlap=sign_table["TRUE","TRUE"],
      same_direction = sign_table["TRUE","TRUE"] + sign_table["FALSE","FALSE"],
      opposite_direction = sign_table["TRUE","FALSE"] + sign_table["FALSE","TRUE"],
      p=fisher.test(sign_table,alt="l")$p.value))
  }else{
    dataset_direction_stats = rbind(dataset_direction_stats,
      data.frame(dataset=datasetname,test = "sign_test_female",tissue=tissue,
      bg_size = sum(sign_table),overlap=0,
      same_direction = 0,
      opposite_direction = 0,
      p=NA))
  }
  
  assign("dataset_direction_stats", dataset_direction_stats, envir = .GlobalEnv)
  
  selected_sets[["female_up_disease_down"]] = shared_genes[
    fc1[shared_genes] < 0 & fc2[shared_genes] > 0]
  selected_sets[["female_down_disease_up"]]  = shared_genes[
    fc1[shared_genes] > 0 & fc2[shared_genes] < 0]
  selected_sets[["female_down_disease_down"]]  = shared_genes[
    fc1[shared_genes] < 0 & fc2[shared_genes] < 0]
  selected_sets[["female_up_disease_up"]]  = shared_genes[
    fc1[shared_genes] < 0 & fc2[shared_genes] < 0]
  
  dataset2selected_sets[[datasetname]] = selected_sets
  assign("dataset2selected_sets", dataset2selected_sets, envir = .GlobalEnv)
  
  # Enrichment analysis using GO
  res = gost(selected_sets,
      organism=enrichment_species,
      significant=FALSE, # return all results
      correction_method='fdr', # it will not return unadjusted p-values. 
      sources=c('KEGG','REAC','WP'),
      custom_bg = bg,
      domain_scope='custom',
      evcodes=T)
  res$result$q_value = p.adjust(res$result$p_value,method="fdr")
  res$result = res$result[res$result$term_size < 200,]
  res$result = res$result[order(res$result$q_value),]
  dataset2set_enrichment_analysis[[datasetname]] = res$result
  assign("dataset2set_enrichment_analysis", dataset2set_enrichment_analysis, envir = .GlobalEnv)
  return(NULL)
}

```

Get feature mappings and mitocarta annotation

```{r}
feature2gene = FEATURE_TO_GENE[!grepl("chr\\d",FEATURE_TO_GENE$feature_ID),]
feature2gene = feature2gene[!grepl("cluster\\d",feature2gene$feature_ID),]
feature2gene = feature2gene[!grepl("chrX",feature2gene$feature_ID),]
feature2gene = merge(feature2gene,RAT_TO_HUMAN_GENE,by.x = "gene_symbol",by.y = "RAT_SYMBOL")
feature2human_gene = split(feature2gene$HUMAN_ORTHOLOG_SYMBOL,feature2gene$feature_ID)
feature2human_gene = lapply(feature2human_gene, unique)

mt_pw = fread(paste0(data_dir,"Human.MitoCarta3.0.txt"),header = T,
        stringsAsFactors = F,data.table = F)
mt_genes = unique(unlist(strsplit(mt_pw$Genes,split=",\\s+")))
mitocarta_pathways = strsplit(mt_pw$Genes,split=",\\s+")
names(mitocarta_pathways) = mt_pw$MitoPathway

# Get uniprot to gene name mapping
library(org.Mm.eg.db)
list_to_mat<-function(l){
  l = as.list(l)
  lengths = sapply(l,length)
  l = l[lengths>0]
  lengths = lengths[lengths>0]
  v = unlist(l)
  s = rep(names(l),lengths)
  d = unique(as.data.frame(cbind(v,s),stringsAsFactors=F))
  rownames(d) = NULL
  return(d)
}
enz_uni = list_to_mat(org.Mm.egUNIPROT)
names(enz_uni) = c("uniprot","entrez")
enz_uni = enz_uni[!is.na(enz_uni$uniprot),]
enz_symb = list_to_mat(org.Mm.egALIAS2EG)
names(enz_symb) = c("entrez","symbol")
uni2symb = merge(enz_uni,enz_symb,by="entrez")
uni2symb = unique(uni2symb[,c("uniprot","symbol")])
dim(uni2symb)
# same for human
library(org.Hs.eg.db)
enz_uni_human = list_to_mat(org.Hs.egUNIPROT)
names(enz_uni_human) = c("uniprot","entrez")
enz_uni_human = enz_uni_human[!is.na(enz_uni_human$uniprot),]
enz_symb_human = list_to_mat(org.Hs.egALIAS2EG)
names(enz_symb_human) = c("entrez","symbol")
uni2symb_human = merge(enz_uni_human,enz_symb_human,by="entrez")
uni2symb_human = unique(uni2symb_human[,c("uniprot","symbol")])
dim(uni2symb_human)
```

Define the stats to keep over the datasets:

```{r}
dataset_overlap_stats = c()
dataset_direction_stats = c()
dataset2selected_sets = list()
dataset2set_enrichment_analysis = list()
```

# Ohman et al.: proteomics T2D in muscle

```{r}
ohman_d = fread(file=paste0(disease_datasets_dir,"ohman_2021_iscience.csv"),
               data.table = F,stringsAsFactors = F)
ohman_d$t2d_fc = log2(ohman_d$SWATH_T2D/ohman_d$SWATH_NGT)

motrpac_d = PROT_SKMGN_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  ohman_d,motrpac_d,
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "Gene name",
  dataset_qval_col = "ANOVA_q-value",
  dataset_fc_col = "t2d_fc",
  datasetname = "T2D,Ohman et al.",
  tissue = "SKMGN"
)


# mito_en_res = lapply(selected_sets, 
#     simple_enrichment_analysis,
#     s = selected_sets[[1]],
#     univ = shared_genes,
#     pathways = mitocarta_pathways,
#     min_size = 3
# )
# for(j in 1:length(mito_en_res)){mito_en_res[[j]]$setname=names(mito_en_res)[j]}
# mito_en_res = do.call(rbind,mito_en_res)
# mito_en_res$q = p.adjust(mito_en_res$p,method="fdr")
# mito_en_res = mito_en_res[order(mito_en_res$p),]
# print(mito_en_res[1:5,])

# free space
rm(ohman_d);rm(motrpac_d)

```

# Chae et al.: proteomics T2D in muscle

```{r}
chae_d = fread(file=paste0(disease_datasets_dir,"chae_2018_emm_stat.csv"),
               data.table = F,stringsAsFactors = F)
chae_d$t2d_fc = chae_d$`log2-Fold change (T2DM / NGT)`

# get the bg gene set
chae_d_set = fread(file=paste0(disease_datasets_dir,"chae_2018_emm_set.csv"),
               data.table = F,stringsAsFactors = F)
chae_d_uni_ids = sapply(
  strsplit(chae_d_set$`UniProt ID`,split="\\%"),
  function(x){
    arr = strsplit(x,split="\\|")[[1]]
    arr[seq(from=2,to=length(arr),by=2)]
  }
)
chae_d_gene_set = merge(
  data.frame(uniprot = chae_d_uni_ids),
  uni2symb_human,by="uniprot"
)

motrpac_d = PROT_SKMGN_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  chae_d,motrpac_d,
  dataset_d_gene_set = unique(chae_d_gene_set$symbol),
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "Symbol",
  dataset_qval_col = "FDR",
  dataset_fc_col = "t2d_fc",
  datasetname = "T2D,Chae et al.",
  tissue = "SKMGN"
)

# free space
rm(chae_d);rm(motrpac_d)
```

<!-- # Ubaida et al.: proteomics, muscle, aging -->

<!-- ```{r} -->
<!-- ubaida_d = fread(file=paste0(disease_datasets_dir,"ubaida_mohien_ 2019_elife_stat.csv"), -->
<!--                data.table = F,stringsAsFactors = F) -->
<!-- ubaida_d_genes = strsplit(ubaida_d$Genenames,split=",|\\s+") -->
<!-- ubaida_d_genes = cbind( -->
<!--   rep(ubaida_d$GenePrimary,sapply(ubaida_d_genes,length)), -->
<!--   unname(unlist(ubaida_d_genes)) -->
<!-- ) -->
<!-- colnames(ubaida_d_genes) = c("GenePrimary","symbol") -->
<!-- ubaida_d = merge(ubaida_d,ubaida_d_genes,by="GenePrimary") -->
<!-- ubaida_d$fc = ubaida_d$AgeBeta -->
<!-- ubaida_d$q = p.adjust(ubaida_d$`BH Adjusted Pvalue`) -->

<!-- ubaida_set = fread(file=paste0(disease_datasets_dir,"uniprot2gene_name_ubaida.tsv"), -->
<!--                data.table = F,stringsAsFactors = F) -->

<!-- motrpac_d = PROT_SKMGN_DA -->
<!-- motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID") -->

<!-- add_overlap_tests_and_enrichments( -->
<!--   ubaida_d,motrpac_d, -->
<!--   dataset_d_gene_set = unique(ubaida_set$To), -->
<!--   motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL", -->
<!--   dataset_gene_col = "symbol", -->
<!--   dataset_qval_col = "q", -->
<!--   dataset_fc_col = "fc", -->
<!--   datasetname = "Aging,Ubaida et al.", -->
<!--   tissue = "SKMGN" -->
<!-- ) -->

<!-- # free space -->
<!-- rm(ubaida_d);rm(motrpac_d) -->
<!-- ``` -->

# Stocks: proteomics in liver, obesity

```{r}
stocks_d = fread(file=paste0(disease_datasets_dir,"stocks_2022_molcellproteomics_stat.csv"),
               data.table = F,stringsAsFactors = F)

l = strsplit(stocks_d$`Gene names`,split=";")
names(l) = stocks_d$`Protein IDs`
stocks_gene_mat = unique(cbind(
  rep(names(l),sapply(l,length)),
  unname(unlist(l))
))
stocks_gene_mat = as.data.frame(stocks_gene_mat)
names(stocks_gene_mat) = c("id","symbol")
stocks_d = merge(stocks_d,stocks_gene_mat,by.x = "Protein IDs",by.y="id")
stocks_d$ob_fc = stocks_d$`Student's T-test Test statistic ob_lean`
stocks_d$q = stocks_d$`Student's T-test q-value ob_lean`

motrpac_d = PROT_LIVER_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  stocks_d,motrpac_d,
  motrpac_gene_col = "gene_symbol",
  dataset_gene_col = "symbol",
  dataset_qval_col = "q",
  dataset_fc_col = "ob_fc",
  datasetname = "Obesity,Stocks et al.",
  tissue = "LIVER",
  enrichment_species = "rnorvegicus"
)

# free space
rm(stocks_d);rm(motrpac_d)
```


<!-- # Zhang.: proteomics in liver, high fat diet in rat -->


<!-- ```{r} -->
<!-- zhang_d = fread(file=paste0(disease_datasets_dir,"zhang_2010_hepatology.csv"), -->
<!--                data.table = F,stringsAsFactors = F) -->

<!-- ipi_db = fread(paste0(data_dir,"ipi.genes.RAT.xrefs"),data.table = F,stringsAsFactors = F) -->
<!-- gene2ipi = strsplit(ipi_db$V10,split=";") -->
<!-- names(gene2ipi) = ipi_db$V7 -->
<!-- gene_ipi_mat = unique(cbind( -->
<!--   rep(names(gene2ipi),sapply(gene2ipi,length)), -->
<!--   unname(unlist(gene2ipi)) -->
<!-- )) -->
<!-- gene_ipi_mat = as.data.frame(gene_ipi_mat) -->
<!-- names(gene_ipi_mat) = c("entrez","ipi") -->
<!-- gene_ipi_mat = merge(gene_ipi_mat,feature2gene[,c("feature_ID","gene_symbol")], -->
<!--                      by.x = "entrez",by.y ="feature_ID") -->
<!-- zhang_d = merge(zhang_d,gene_ipi_mat,by.x = "Accession Number",by.y="ipi") -->
<!-- zhang_d$hfd_fc = log2(as.numeric(zhang_d$HFD_vs_control_ratio_24w)) -->
<!-- zhang_d$q = zhang_d$HFD_vs_control_pval_24w -->

<!-- motrpac_d = PROT_LIVER_DA -->
<!-- motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID") -->

<!-- add_overlap_tests_and_enrichments( -->
<!--   zhang_d,motrpac_d, -->
<!--   motrpac_gene_col = "gene_symbol", -->
<!--   dataset_gene_col = "gene_symbol", -->
<!--   dataset_qval_col = "q", -->
<!--   dataset_fc_col = "hfd_fc", -->
<!--   datasetname = "HFD,Zhang et al.", -->
<!--   tissue = "LIVER", -->
<!--   enrichment_species = "rnorvegicus" -->
<!-- ) -->

<!-- # free space -->
<!-- rm(zhang_d);rm(motrpac_d) -->
<!-- ``` -->

# Yuan et al.: proteomics in liver

As of Oct 2022, the full dataset is not available, as mentioned in the ProteomeXchange website:"ProteomeXchange dataset PXD016405 has been reserved by the PRIDE repository for a dataset that has been deposited, but is not yet publicly".

```{r}
yuan_d = fread(file=paste0(disease_datasets_dir,"yuan_2020_jproteomics.csv"),
               data.table = F,stringsAsFactors = F)
yuan_d$nafld_fc = log2(yuan_d$`Fold changed (NAFLD/MHO)`)
yuan_d$q = yuan_d$`p value`

motrpac_d = PROT_LIVER_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  yuan_d,motrpac_d,
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "GENE name",
  dataset_qval_col = "q",
  dataset_fc_col = "nafld_fc",
  datasetname = "NAFLD,Yuan et al.",
  tissue = "LIVER"
)

# free space
rm(yuan_d);rm(motrpac_d)
```

# Niu et al.: LIVER, Cirrhosis, NASH proteomics

```{r}
niue_dataset = fread(file=paste0(disease_datasets_dir,"niu_2022_molsystbiol_set.csv"),
               data.table = F,stringsAsFactors = F)
niu_genes = strsplit(niue_dataset$PG.Genes,split=";")
names(niu_genes) = niue_dataset$PG.Genes
niu_acc = strsplit(niue_dataset$PG.ProteinAccessions,split=";")
names(niu_acc) = niue_dataset$PG.ProteinAccessions
niu_prot_genes = c()
for(i in 1:length(niu_genes)){
  gs = niu_genes[[i]]
  ps = niu_acc[[i]]
  for(g in gs){
    for(p in ps){
      niu_prot_genes = rbind(niu_prot_genes,c(g,p,names(niu_genes)[i],names(niu_acc)[i]))
    }
  }
}
niu_prot_genes = as.data.frame(niu_prot_genes,stringsAsFactors = F)
names(niu_prot_genes) = c("gene","prot","genes","prots")
niu_prot_genes = unique(niu_prot_genes)
dim(niu_prot_genes)
nash_vs_h_fcs = rowMeans(log2(niue_dataset[,colnames(niue_dataset)=="NASH"])) - 
  rowMeans(log2(niue_dataset[,colnames(niue_dataset)=="Control"]))
cir_vs_h_fc = rowMeans(log2(niue_dataset[,colnames(niue_dataset)=="Cirrhosis"])) - 
  rowMeans(log2(niue_dataset[,colnames(niue_dataset)=="Control"]))
nash_vs_h_fcs = cbind(niue_dataset$PG.ProteinAccessions,nash_vs_h_fcs,cir_vs_h_fc)
colnames(nash_vs_h_fcs) = c("prots","nash_fc","cir_fc")
niu_prot_genes = merge(niu_prot_genes,nash_vs_h_fcs,by="prots")

niu_d = fread(file=paste0(disease_datasets_dir,"niu_2022_molsystbiol_stat.csv"),
               data.table = F,stringsAsFactors = F)
niu_d = merge(niu_d,unique(niu_prot_genes[,c("gene","nash_fc","cir_fc")]),by.x="Gene name",by.y="gene")

motrpac_d = PROT_LIVER_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

niu_d$nash_q = 1-as.numeric(niu_d$`Post-hoc Tukey NASH vs. Healthy`=="Significant")
niu_d$cirr_q = 1-as.numeric(niu_d$`Post-hoc Tukey Cirrhosis vs. Healthy`=="Significant")
niu_d$nash_fc = as.numeric(as.character(niu_d$nash_fc))
niu_d$cir_fc = as.numeric(as.character(niu_d$cir_fc))

add_overlap_tests_and_enrichments(
  niu_d,motrpac_d,
  dataset_d_gene_set = unique(niu_prot_genes$gene),
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "Gene name",
  dataset_qval_col = "nash_q",
  dataset_fc_col = "nash_fc",
  datasetname = "NASH,Niu et al.",
  tissue = "LIVER"
)
add_overlap_tests_and_enrichments(
  niu_d,motrpac_d,
  dataset_d_gene_set = unique(niu_prot_genes$gene),
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "Gene name",
  dataset_qval_col = "cirr_q",
  dataset_fc_col = "cir_fc",
  datasetname = "Cirr,Niu et al.",
  tissue = "LIVER"
)


```


# Park et al.: Heart, MI

```{r}
park_d = fread(file=paste0(disease_datasets_dir,"park_2019_celldeathdis.csv"),
               data.table = F,stringsAsFactors = F)
park_d$mi_fc = log2(park_d$`FC (MI 8W/Sham)`)
park_d$p = park_d$`p-value`
park_d$q = p.adjust(park_d$p,method="fdr")
# merge data using the uniprot mouse mapping from ncbi
park_d = merge(park_d,uni2symb,by.x="Accession",by.y="uniprot")
park_d = park_d[!is.na(park_d$symbol),]

motrpac_d = PROT_HEART_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  park_d,motrpac_d,
  motrpac_gene_col = "gene_symbol",
  dataset_gene_col = "symbol",
  dataset_qval_col = "q",
  dataset_fc_col = "mi_fc",
  datasetname = "MI,Park et al.",
  tissue = "HEART",
  enrichment_species = "rnorvegicus"
)

# free space
rm(park_d);rm(motrpac_d)
```

# Coats et al.: Heart HCM

```{r}
coats_d = fread(file=paste0(disease_datasets_dir,"coats_2018_circgpm_stat.csv"),
               data.table = F,stringsAsFactors = F)
coats_gene_map = fread(file=paste0(disease_datasets_dir,"uniport2gene_name_coats.tsv"),
               data.table = F,stringsAsFactors = F)
names(coats_gene_map)[2] = "symbol"
coats_d = merge(coats_d,coats_gene_map,by.x="UniProt Accession",by.y="From")
coats_d$hcm_fc = coats_d$`Max fold change`
coats_d$p = coats_d$`Anova (p)`
coats_d$q = coats_d$`q Value`

motrpac_d = PROT_HEART_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

coats_gene_bg = fread(file=paste0(disease_datasets_dir,"uniprot2gene_name_coats_bg.tsv"),
               data.table = F,stringsAsFactors = F)


add_overlap_tests_and_enrichments(
  coats_d,motrpac_d,
  dataset_d_gene_set = unique(coats_gene_bg$To),
  motrpac_gene_col = "HUMAN_ORTHOLOG_SYMBOL",
  dataset_gene_col = "symbol",
  dataset_qval_col = "q",
  dataset_fc_col = "hcm_fc",
  datasetname = "HCM,Coats et al.",
  tissue = "HEART"
)

# free space
rm(coats_d);rm(motrpac_d)
```

# Havlenova, Heart, HF

```{r}
hav_d = fread(file=paste0(disease_datasets_dir,"havlenova_2021_scirep.csv"),
               data.table = F,stringsAsFactors = F)
hav_d$hf_fc = log2(hav_d$`RV Fold-change`)
hav_d$q  = hav_d$`RV Adj. P-Value`
hav_d$q[hav_d$q=="<0.01"]="0.01"
hav_d$q = as.numeric(hav_d$q)

motrpac_d = PROT_HEART_DA
motrpac_d = merge(motrpac_d,feature2gene,by="feature_ID")

add_overlap_tests_and_enrichments(
  hav_d,motrpac_d,
  motrpac_gene_col = "gene_symbol",
  dataset_gene_col = "Gene Symbol",
  dataset_qval_col = "q",
  dataset_fc_col = "hf_fc",
  datasetname = "HF,Havlenova et al.",
  tissue = "HEART",
  enrichment_species = "rnorvegicus"
)
```

# Visualizations of the results 


```{r}

# Write the stats tables for supplement
write.table(
  dataset_overlap_stats,
  sep="\t",
  quote=F,
  row.names = F,
  file="~/Desktop/repos/motrpac-rat-training-mitochondria/data/human_data_comparison_overlap_stats.tsv"
)
write.table(
  dataset_direction_stats,
  sep="\t",
  quote=F,
  row.names = F,
  file="~/Desktop/repos/motrpac-rat-training-mitochondria/data/human_data_comparison_direction_stats.tsv"
)

dataset_overlap_stats$sex = sapply(dataset_overlap_stats$test,
    function(x){arr=strsplit(x,split="_")[[1]];arr[length(arr)]})
dataset_overlap_stats$log10P = -log10(dataset_overlap_stats$p)

dataset_overlap_stats$name = paste(dataset_overlap_stats$tissue,
      dataset_overlap_stats$dataset,sep=":")
dataset_overlap_stats$name2 = paste(dataset_overlap_stats$tissue,
      dataset_overlap_stats$dataset,dataset_overlap_stats$sex,sep=",")

dataset_overlap_stats$tissue[
  dataset_overlap_stats$tissue=="SKMGN"
] = "SKM-GN"

# par(mar=c(12,5,2,2))
# currinds = dataset_overlap_stats$sex=="p"
# barplot(
#   dataset_overlap_stats[currinds,"log10P"],
#   col = TISSUE_COLORS[dataset_overlap_stats$tissue[currinds]],
#   names.arg = gsub(" et al.","",dataset_overlap_stats[currinds,"name"]),
#   ylab = "Overlap -log10 p-value",las=2
# );abline(h=-log10(0.05),lty=2,col="red")
# 
# par(mar=c(12,5,2,2))
# currinds = dataset_overlap_stats$sex!="p"
# barplot(
#   dataset_overlap_stats[currinds,"log10P"],
#   col = TISSUE_COLORS[dataset_overlap_stats$tissue[currinds]],
#   las=2,
#   names.arg = gsub(" et al.","",dataset_overlap_stats[dataset_overlap_stats$sex!="p","name2"]),
#   ylab = "Discordant FC sign -log10 p-value"
# );abline(h=-log10(0.05),lty=2,col="red")

# Use ggplot
library(ggplot2)
df = dataset_overlap_stats
df$Study = gsub(" et al.","",df$name)
df[df$sex == "p","sex"] = "Overall set overlap"
df[df$sex == "male","sex"] = "Male:opposite FC"
df[df$sex == "female","sex"] = "Female:opposite FC"
df$Weight = format(df$p,digits = 2)
df$Weight[df$log10P < 1.5] = NA
df$Test = df$sex
df$logPthreshold = df$log10P > 5
df$log10P = pmin(5,df$log10P)

# ggplot(df,aes(x = Test,y = log10P,fill = Study,color=Study)) +
#   geom_bar(stat = "identity",position = "dodge") + 
#   theme(axis.text.x = element_text(angle = 45,vjust = 1, hjust = 1))  +
#   geom_hline(yintercept = -log10(0.05)) + ylab("-log10 p-value") + 
#   geom_text(aes(x = Test,y = log10P+0.3,label = Weight),
#             vjust = 1.5, position = position_dodge(.9),
#             col="black",size=3,angle=45)

# ggplot(df[!grepl("Overall",df$sex),],
#   aes(x = Test,y = log10P,fill = Study,color=Study)) +
#   geom_bar(stat = "identity",position = "dodge") + 
#   theme(axis.text.x = element_text(angle = 45,vjust = 1, hjust = 1))  +
#   geom_hline(yintercept = -log10(0.05)) + ylab("-log10 p-value") + 
#   geom_text(aes(x = Test,y = log10P+0.3,label = Weight),
#             vjust = 1.5, position = position_dodge(.9),
#             col="black",size=3,angle=45)

# ggplot(df[grepl("Overall",df$sex),],
#   aes(x = Test,y = log10P,fill = Study,color=Study)) +
#   geom_bar(stat = "identity",position = "dodge") + 
#   theme(axis.text.x = element_blank())  + xlab("Overall set overlap") + 
#   geom_hline(yintercept = -log10(0.05)) + ylab("-log10 p-value") + 
#   geom_text(aes(x = Test,y = log10P+0.3,label = Weight),
#             vjust = 1.5, position = position_dodge(.9),
#             col="black",size=3,angle=45)

df$name3 = sapply(df$name2,function(x)paste(strsplit(x,split=",")[[1]][-1],collapse=","))
df_overall_test = df[grepl("Overall",df$sex),]
df_overall_test = df_overall_test[order(df_overall_test$tissue),]
df_overall_test$name3 = factor(df_overall_test$name3,
                               levels = df_overall_test$name3)
df_overall_test$name4 = gsub(" et al.,p","",df_overall_test$name3)
df_overall_test$Tissue = df_overall_test$tissue
ggplot(df_overall_test,
  aes(x = name4,y = log10P,fill = Tissue)) +
  geom_bar(stat = "identity",position = "dodge") + xlab("") + 
  geom_hline(yintercept = -log10(0.05)) + ylab("-log10 p-value") + 
  geom_text(aes(x = name4,y = log10P-0.25,label = Weight),
            vjust = 0.5, position = position_dodge(.5),
            col="black",size=4.8,angle=45) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1,size=13)) +
  theme(axis.text.y = element_text(size=12),
        axis.title=element_text(size=14,face="bold"))

df_overall_test[,c("dataset","tissue","bg_size","overlap","p")]


dataset_direction_stats$log10P = -log10(dataset_direction_stats$p)
dataset_direction_stats$sex = "male"
dataset_direction_stats$sex[grepl("female",dataset_direction_stats$test)] = "female"
dataset_direction_stats$Weight = format(dataset_direction_stats$p,digits = 2)
dataset_direction_stats$Weight[dataset_direction_stats$log10P < 1.5] = NA
dataset_direction_stats$logPthreshold = dataset_direction_stats$log10P > 5
dataset_direction_stats$log10P = pmin(5,dataset_direction_stats$log10P)

dataset_direction_stats = dataset_direction_stats[
  order(
    dataset_direction_stats$tissue,decreasing = T
  ),
]
dataset_direction_stats$Tissue = dataset_direction_stats$tissue
dataset_direction_stats$Dname = gsub(" et al.","",dataset_direction_stats$dataset)
ggplot(dataset_direction_stats,
  aes(x = Dname,y = log10P,fill = Tissue)) +
  facet_wrap(~sex,  ncol=1) +
  geom_bar(stat = "identity",position = "dodge") + xlab("") + 
  theme_classic() + 
  theme(axis.text.x = element_text(angle = 60,vjust = 1, hjust = 1,size=14),
        strip.background = element_rect(colour="black",fill="cyan"),
        strip.text = element_text(size = 15),
        axis.title=element_text(size=14,face="bold"))  +
  geom_hline(yintercept = -log10(0.05),color="gray") + ylab("-log10 p-value") + 
  geom_text(aes(x = Dname,y = log10P+0.3,label = Weight),
            vjust = 1.5, position = position_dodge(.9),
            col="black",size=3.8,angle=35)

```

```{r}
selected_enr_res = lapply(dataset2set_enrichment_analysis,function(x)x[x$q_value<0.1,])
selected_enr_res = lapply(selected_enr_res,function(x)x[x$intersection_size>2,])
selected_enr_res = selected_enr_res[sapply(selected_enr_res,nrow)>0]
for(nn in names(selected_enr_res)){selected_enr_res[[nn]]$dataset=nn}
selected_enr_res = do.call(rbind,selected_enr_res)
selected_enr_res = selected_enr_res[,c("dataset","query","p_value","q_value","term_size",
                                       "query_size","intersection_size","term_name",
                                       "effective_domain_size","intersection")]
write.table(
  selected_enr_res,
  sep="\t",
  quote=F,
  row.names = F,
  file="~/Desktop/repos/motrpac-rat-training-mitochondria/data/human_data_comparison_enrichments.tsv"
)
```

Create gene set files for cytoscape

```{r}
cirr_sets = dataset2selected_sets$`Cirr,Niu et al.`
cirr_genes = data.frame(
  gene = cirr_sets$female_up_disease_down,
  direction = "exercise_up"
)
cirr_genes = rbind(cirr_genes,
  data.frame(
    gene = cirr_sets$female_down_disease_up,
    direction = "exercise_down"
  )
)
write.table(
  cirr_genes,
  sep = "\t", 
  file = paste0(data_dir,"/selected_cirr_gene_set.tsv"),
  quote=F,row.names=F
)

diab_sets1 = dataset2selected_sets$`T2D,Ohman et al.`
diab_sets2 = dataset2selected_sets$`T2D,Chae et al.`
diab_sets = list()
for(s in names(diab_sets1)){
  diab_sets[[s]] = intersect(diab_sets1[[s]],diab_sets2[[s]])
}
sapply(diab_sets, length)
diab_genes = data.frame(
  gene = intersect(diab_sets$male_up_disease_down,diab_sets$female_up_disease_down),
  direction = "exercise_up"
)
diab_genes = rbind(
  diab_genes,
  data.frame(
    gene = intersect(diab_sets$male_down_disease_up,diab_sets$female_down_disease_up),
    direction = "exercise_down"
  )
)
write.table(
  diab_genes,
  sep = "\t", 
  file = paste0(data_dir,"/selected_t2d_gene_set.tsv"),
  quote=F,row.names=F
)
```


