---
title: "Mito-adjusted gene expression DEA"
author: "David Amar"
date: "6/21/2022"
output: html_document
---

```{r setup, include=T}
knitr::opts_chunk$set(echo = TRUE)
# load libraries
library(data.table)
library(DESeq2)
library(MotrpacBicQC)
library(ggplot2)
library(metap)
library(IHW)
library(corrplot)
library(MotrpacRatTraining6moData)

# Load useful functions
mawg_repo = "~/Desktop/repos/motrpac-mawg/"
source(paste0(mawg_repo,'pass1b-06/integrative/outliers_covariates/pi1_cook_fx.R'))
source(paste0(mawg_repo,'pass1b-06/tools/get_fx.R'))
#source(paste0(mawg_repo,'pass1b-06/integrative/clustering/cluster_viz_fx.R'))

data_dir = "~/Desktop/repos/motrpac-rat-training-mitochondria/data/"
output_dir = "~/Desktop/MoTrPAC/data/pass1b_6m/"
```


# Load datasets

```{r define data set}
# get animal code data
tissue_codes = names(TISSUE_CODE_TO_ABBREV)

# read the RNA-seq metrics to obtain mtRNA read percentages
rnaseq_qa_qc = TRNSCRPT_META
pct_mito_reads = rnaseq_qa_qc$pct_chrM
names(pct_mito_reads) = as.character(rnaseq_qa_qc$vial_label)
rownames(rnaseq_qa_qc) = as.character(rnaseq_qa_qc$vial_label)
colnames(rnaseq_qa_qc) = tolower(colnames(rnaseq_qa_qc))

# Add outliers and phenotypic data
outliers = as.character(OUTLIERS[,1])
pheno = PHENO
# also add vena cava outliers
venacv_outliers = unique(pheno[
  pheno$sex=="female" & pheno$sacrificetime %in% c("1w","2w") & 
    pheno$specimen.processing.sampletypedescription=="Aorta", "viallabel"])
outliers = union(outliers, venacv_outliers)
rownames(pheno) = as.character(pheno$viallabel)
```

Load mito biomarker data (see `mito_biomarker_analysis.Rmd`)
```{r}
load(paste0(data_dir,"mt_biomarkers.RData"))
```

Load existing mito annotation data (see `mito_biomarker_analysis.Rmd`):
```{r}
load(paste0(data_dir,"mito_gene_annotation.RData"))
```


Define the tissues and features for the adjusted analysis

```{r}
tissues_for_adjustment = c()
adjustment_features = list()
for(tissue in names(mt_biomarkers)){
  if(tissue %in% c("TESTES","OVARY","VENACAVA")){next}
  df = mt_biomarkers[[tissue]]
  df$CL_72_8 = df$`CL(72:8)>CL(18:2/18:2/18:2/18:2)`
  if(all(is.na(df$CL_72_8))){df$CL_72_8=df$`CL(72:8)`}
  if(!all(is.na(df$CL_72_8))){
    tissues_for_adjustment = union(tissues_for_adjustment,tissue)
    adjustment_features[[tissue]] = c("CL_72_8")
  }
  x1 = df$mt_reads
  x2 = df$mtdna
  inds = !is.na(x1) & !is.na(x2)
  if(sum(inds)>10){
    rho = cor(x1[inds],x2[inds])
    if(rho > 0.5){
      tissues_for_adjustment = union(tissues_for_adjustment,tissue)
      adjustment_features[[tissue]] = 
        union(adjustment_features[[tissue]],"mt_reads")
    }
  }
  mt_biomarkers[[tissue]] = df
}
```

# Run differential analysis

See main functions in pass1b-06/tools/get_fx.R. 

## Sanity check: permutation test within sex

Here we permute the training group in each sex and check the number of differential genes found. This is a sanity check for the landscape results.

```{r}
perm_dea_results = list()
for(tissue_name in c("ADRNL","HEART","COLON","SKM-GN")){
  tissue_code = TISSUE_ABBREV_TO_CODE[tissue_name]
  if(tissue_name=="" || tissue_code == ""){next}
  print(paste("Analyzing tissue:",tissue_name))
  
  # load data 
  raw_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_RAW_COUNTS")
  norm_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_NORM_DATA")
  raw_data = get(raw_data_name)
  norm_data = get(norm_data_name)
  rownames(raw_data) = raw_data$feature_ID
  raw_data = raw_data[,-c(1:4)]
  rownames(norm_data) = norm_data$feature_ID
  norm_data = norm_data[,-c(1:4)]
  # filter unexpressed genes by taking the genes in the norm data
  raw_data = raw_data[rownames(norm_data),]
  curr_vials = colnames(norm_data)
  data = list(
    meta = data.table(cbind(pheno[curr_vials,],rnaseq_qa_qc[curr_vials,])),
    counts = raw_data[,curr_vials],
    tmm = norm_data[,curr_vials]
  )
  
  # prep data
  prepped = transcript_prep_data(tissue_code, 
    data$meta, 
    data$counts, 
    data$tmm,
    covariates=c('pct_globin', 'rin', 'pct_umi_dup', 'median_5_3_bias'),
    outliers=outliers)

  # take training groups that appear in both sexes
  reduced_meta = as.data.frame(prepped$fixed_meta)
  rownames(reduced_meta) = reduced_meta$viallabel
  reduced_meta = reduced_meta[!rownames(reduced_meta) %in% outliers,]
  tb = table(reduced_meta$group,reduced_meta$sex)
  tb = tb[!rownames(tb)=="control",]
  if(!is.null(dim(tb)) && ncol(tb)>1){
    tb = tb[apply(tb>2,1,all),]
    training_groups = rownames(tb)
  }else{
    tb = tb[tb>2]
    training_groups = names(tb)
  }
  
  print("analyzing the following timewise contrasts:")
  print(training_groups)
  
  # timewise_dea = transcript_timewise_dea_each_sex(tissue_code, 
  #   prepped$fixed_meta, 
  #   prepped$fixed_counts, 
  #   prepped$fixed_covariates, 
  #   prepped$curr_outliers,
  #   date = "14122022",
  #   training_groups = training_groups,
  #   write=F, save_rdata = F)
  
  training_dea = transcript_training_dea_each_sex(tissue_code, 
    prepped$fixed_meta, 
    prepped$fixed_counts, 
    prepped$fixed_covariates, 
    prepped$curr_outliers,
    date = "14122022",write=F)
  print("Number of DEA at 5% FDR:")
  print(table(p.adjust(training_dea$p_value,method="fdr")<0.05))
  perm_dea_results[[tissue_name]] = list(real = training_dea,perms = list())
  
  print("Number of DEA at 5% FDR after permutation:")
  for(j in 1:10){
    for(sex in unique(prepped$fixed_meta$sex)){
      inds = prepped$fixed_meta$sex == sex
      prepped$fixed_meta$group[inds] = sample(
        prepped$fixed_meta$group[inds],replace = F)
    }
    
    perm_training_dea = transcript_training_dea_each_sex(tissue_code, 
      prepped$fixed_meta, 
      prepped$fixed_counts, 
      prepped$fixed_covariates, 
      prepped$curr_outliers,
      date = "14122022",write=F)
    print(table(p.adjust(perm_training_dea$p_value,method="fdr")<0.05))
    perm_dea_results[[tissue_name]][["perms"]][[j]] = perm_training_dea
  }
}

```

## Sanity check: permutation test for sex differences

Here we permute the sex assignment and rerun the differential analysis pipelines.

```{r}
sex_perm_dea_results = list()
for(tissue_name in sort(unique(names(TISSUE_ABBREV_TO_CODE)))){
  tissue_code = TISSUE_ABBREV_TO_CODE[tissue_name]
  if(tissue_name=="" || tissue_code == ""){next}
  if(tissue_name %in% c("VENACV","TESTES","OVARY","PLASMA")){next}
  print(paste("Analyzing tissue:",tissue_name))
  
  # load data 
  raw_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_RAW_COUNTS")
  norm_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_NORM_DATA")
  raw_data = get(raw_data_name)
  norm_data = get(norm_data_name)
  rownames(raw_data) = raw_data$feature_ID
  raw_data = raw_data[,-c(1:4)]
  rownames(norm_data) = norm_data$feature_ID
  norm_data = norm_data[,-c(1:4)]
  # filter unexpressed genes by taking the genes in the norm data
  raw_data = raw_data[rownames(norm_data),]
  curr_vials = colnames(norm_data)
  data = list(
    meta = data.table(cbind(pheno[curr_vials,],rnaseq_qa_qc[curr_vials,])),
    counts = raw_data[,curr_vials],
    tmm = norm_data[,curr_vials]
  )
  
  # prep data
  prepped = transcript_prep_data(tissue_code, 
    data$meta, 
    data$counts, 
    data$tmm,
    covariates=c('pct_globin', 'rin', 'pct_umi_dup', 'median_5_3_bias'),
    outliers=outliers)

  # take training groups that appear in both sexes
  reduced_meta = as.data.frame(prepped$fixed_meta)
  rownames(reduced_meta) = reduced_meta$viallabel
  reduced_meta = reduced_meta[!rownames(reduced_meta) %in% outliers,]
  tb = table(reduced_meta$group,reduced_meta$sex)
  tb = tb[!rownames(tb)=="control",]
  if(!is.null(dim(tb)) && ncol(tb)>1){
    tb = tb[apply(tb>2,1,all),]
    training_groups = rownames(tb)
  }else{
    tb = tb[tb>2]
    training_groups = names(tb)
  }
  
  print("analyzing the following timewise contrasts:")
  print(training_groups)
  
  timewise_dea = transcript_timewise_dea_each_sex(tissue_code,
    prepped$fixed_meta,
    prepped$fixed_counts,
    prepped$fixed_covariates,
    prepped$curr_outliers,
    date = "14122022",
    training_groups = training_groups,
    write=F, save_rdata = F)
  
  training_dea = transcript_training_dea_each_sex(tissue_code,
    prepped$fixed_meta,
    prepped$fixed_counts,
    prepped$fixed_covariates,
    prepped$curr_outliers,
    date = "14122022",write=F)

  print("Number of DEA at 5% FDR:")
  selected_training_genes = training_dea$feature_ID[p.adjust(training_dea$p_value,method="fdr")<0.05]
  print(length(selected_training_genes))
  
  # Get the logFC and z-score correlations
  logfc_sex_corrs = compute_sex_dea_corr(timewise_dea,selected_training_genes,"logFC")
  zscore_sex_corrs = compute_sex_dea_corr(timewise_dea,selected_training_genes,"zscore")
  real_sex_corrs = data.frame(
    feature_ID = selected_training_genes,
    logfc = logfc_sex_corrs[selected_training_genes],
    zscore = zscore_sex_corrs[selected_training_genes]
  )
  sex_perm_dea_results[[tissue_name]] = list(
    real_sex_corrs = real_sex_corrs,
    perm_sex_corrs = c(),
    perm_selected_genes_sex_corrs = c()
  )
  
  print("Number of DEA at 5% FDR after sex permutation:")
  for(j in 1:10){
    prepped$fixed_meta$sex = sample(prepped$fixed_meta$sex, replace = F)
    
    sex_perm_timewise_dea = transcript_timewise_dea_each_sex(tissue_code,
      prepped$fixed_meta,
      prepped$fixed_counts,
      prepped$fixed_covariates,
      prepped$curr_outliers,
      date = "14122022",
      training_groups = training_groups,
      write=F, save_rdata = F)
    
    sex_perm_training_dea = transcript_training_dea_each_sex(tissue_code, 
      prepped$fixed_meta, 
      prepped$fixed_counts, 
      prepped$fixed_covariates, 
      prepped$curr_outliers,
      date = "14122022",write=F)
    
    print(table(p.adjust(sex_perm_training_dea$p_value,method="fdr")<0.05))
    sex_perm_selected_training_genes = sex_perm_training_dea$feature_ID[
      p.adjust(sex_perm_training_dea$p_value,method="fdr")<0.05]
    print(length(sex_perm_selected_training_genes))
    print(paste("Overlap of perm and real:",length(intersect(
      sex_perm_selected_training_genes,selected_training_genes
    ))))
    
    curr_logfc_sex_corrs = compute_sex_dea_corr(sex_perm_timewise_dea,selected_training_genes,"logFC")
    curr_zscore_sex_corrs = compute_sex_dea_corr(sex_perm_timewise_dea,selected_training_genes,"zscore")
    perm_sex_corrs = data.frame(
      feature_ID = selected_training_genes,
      logfc = curr_logfc_sex_corrs[selected_training_genes],
      zscore = curr_zscore_sex_corrs[selected_training_genes],
      rep = j
    )
    curr_logfc_sex_corrs = compute_sex_dea_corr(
      sex_perm_timewise_dea,sex_perm_selected_training_genes,"logFC")
    curr_zscore_sex_corrs = compute_sex_dea_corr(
      sex_perm_timewise_dea,sex_perm_selected_training_genes,"zscore")
    perm_sex_corrs2 = data.frame(
      feature_ID = sex_perm_selected_training_genes,
      logfc = curr_logfc_sex_corrs[sex_perm_selected_training_genes],
      zscore = curr_zscore_sex_corrs[sex_perm_selected_training_genes],
      rep = j
    )
    sex_perm_dea_results[[tissue_name]]$perm_sex_corrs = rbind(
      sex_perm_dea_results[[tissue_name]]$perm_sex_corrs,
      perm_sex_corrs
    )
    sex_perm_dea_results[[tissue_name]]$perm_selected_genes_sex_corrs = rbind(
      sex_perm_dea_results[[tissue_name]]$perm_selected_genes_sex_corrs,
      perm_sex_corrs2
    )
    
    # plot the two distributions
    hist_df = data.frame(
      rho = c(real_sex_corrs$logfc,perm_sex_corrs$logfc),
      var = c(rep("real",nrow(real_sex_corrs)),rep("perm",nrow(perm_sex_corrs)))
    )
    p = ggplot(hist_df, aes(x=rho, fill=var)) +
      geom_histogram( color='#e9ecef', alpha=0.6, position='identity') + 
      ggtitle(paste(tissue_name,", permutation repeat:",j))
    print(p)
    print(paste(
      "Real vs. perm correlations, paired Wilcox test p:",
      wilcox.test(
        real_sex_corrs$logfc,
        perm_sex_corrs$logfc,
        paired = T
      )$p.value))
  }
}

library(tidyr)
compute_sex_dea_corr<-function(timewise,ids,column_name = "logFC",add_zero=T){
  timewise = as.data.frame(timewise)
  timewise = timewise[timewise$feature_ID %in% ids,]
  timewise_male = timewise[timewise$sex=="male",]
  timewise_female = timewise[timewise$sex=="female",]
  timewise_male = timewise_male[,c("feature_ID","comparison_group",column_name)]
  timewise_female = timewise_female[,c("feature_ID","comparison_group",column_name)]
  timewise_male = reshape(timewise_male,idvar="feature_ID",timevar = "comparison_group",direction = "wide")
  timewise_female = reshape(timewise_female,idvar="feature_ID",timevar = "comparison_group",direction = "wide")
  rownames(timewise_male) = timewise_male$feature_ID
  rownames(timewise_female) = timewise_female$feature_ID
  timewise_male = timewise_male[,-1]
  timewise_female = timewise_female[,-1]
  timewise_female = timewise_female[rownames(timewise_male),colnames(timewise_female)]
  if(add_zero){
    timewise_male = cbind(0,timewise_male)
    timewise_female = cbind(0,timewise_female)
  }
  sex_corrs = diag(cor(
    t(timewise_male),
    t(timewise_female)
  ))
  return(sex_corrs)
}

```

## Adjust for mtDNA levels

```{r rna-seq dea table,eval=T}
dea_results = list()
for(tissue_name in unique(pheno$tissue)){
  tissue_code = TISSUE_ABBREV_TO_CODE[tissue_name]
  if(tissue_name=="" || tissue_code == ""){next}
  if(!gsub("-","",tissue_name) %in% names(mt_biomarkers)){next}
  if(!gsub("-","",tissue_name) %in% tissues_for_adjustment){next}
  if(tissue_code %in% names(dea_results)){next}
  
  # define the columns for the current adjustment analysis
  adj_mt_cols = adjustment_features[[gsub("-","",tissue_name)]]

  # load data 
  raw_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_RAW_COUNTS")
  norm_data_name = paste0("TRNSCRPT_",gsub("-","",tissue_name),"_NORM_DATA")
  raw_data = get(raw_data_name)
  norm_data = get(norm_data_name)
  rownames(raw_data) = raw_data$feature_ID
  raw_data = raw_data[,-c(1:4)]
  rownames(norm_data) = norm_data$feature_ID
  norm_data = norm_data[,-c(1:4)]
  # filter unexpressed genes by taking the genes in the norm data
  raw_data = raw_data[rownames(norm_data),]
  curr_vials = colnames(norm_data)
  data = list(
    meta = data.table(cbind(pheno[curr_vials,],rnaseq_qa_qc[curr_vials,])),
    counts = raw_data[,curr_vials],
    tmm = norm_data[,curr_vials]
  )
  
  # prep data
  prepped = transcript_prep_data(tissue_code, 
    data$meta, 
    data$counts, 
    data$tmm,
    covariates=c('pct_globin', 'rin', 'pct_umi_dup', 'median_5_3_bias'),
    outliers=outliers)
  
  curr_mt_biomarkers = mt_biomarkers[[gsub("-","",tissue_name)]][adj_mt_cols]
  biomarkers_to_keep = colSums(is.na(curr_mt_biomarkers))==0
  shared_vials = intersect(
    rownames(curr_mt_biomarkers),
    prepped$fixed_meta$viallabel
  )
  prepped$fixed_meta = as.data.frame(prepped$fixed_meta)
  rownames(prepped$fixed_meta) = as.character(prepped$fixed_meta$viallabel)
  prepped$fixed_meta = prepped$fixed_meta[shared_vials,]
  prepped$fixed_counts = prepped$fixed_counts[,shared_vials]
  for(newcol in names(which(biomarkers_to_keep))){
    prepped$fixed_meta[[newcol]] = curr_mt_biomarkers[shared_vials,newcol]
  }
  prepped$fixed_meta = data.table(prepped$fixed_meta)
  
  # take training groups that appear in both sexes
  reduced_meta = as.data.frame(prepped$fixed_meta)
  rownames(reduced_meta) = reduced_meta$viallabel
  reduced_meta = reduced_meta[!rownames(reduced_meta) %in% outliers,]
  tb = table(reduced_meta$group,reduced_meta$sex)
  tb = tb[!rownames(tb)=="control",]
  if(!is.null(dim(tb)) && ncol(tb)>1){
    tb = tb[apply(tb>2,1,all),]
    training_groups = rownames(tb)
  }
  else{
    tb = tb[tb>2]
    training_groups = names(tb)
  }
  
  print("analyzing the following timewise contrasts:")
  print(training_groups)
  
  timewise_dea = transcript_timewise_dea_each_sex(tissue_code, 
    prepped$fixed_meta, 
    prepped$fixed_counts, 
    prepped$fixed_covariates, 
    prepped$curr_outliers,
    date = "17082022",
    training_groups = training_groups,
    write=F, save_rdata = F)
  
  training_dea = transcript_training_dea_each_sex(tissue_code, 
    prepped$fixed_meta, 
    prepped$fixed_counts, 
    prepped$fixed_covariates, 
    prepped$curr_outliers,
    date = "17082022",write=F)
  
  prepped$fixed_covariates = c(prepped$fixed_covariates,
                               names(which(biomarkers_to_keep)))
  
  timewise_dea_mt_adjusted = transcript_timewise_dea_each_sex(tissue_code, 
    prepped$fixed_meta, 
    prepped$fixed_counts, 
    prepped$fixed_covariates, 
    prepped$curr_outliers,
    date = "17082022",
    training_groups = training_groups,
    write=F, save_rdata = F)
  
  training_dea_mt_adjusted = transcript_training_dea_each_sex(tissue_code, 
    prepped$fixed_meta, 
    prepped$fixed_counts, 
    prepped$fixed_covariates, 
    prepped$curr_outliers,
    date = "17082022",write=F)
  
  dea_results[[gsub("-","",tissue_name)]] = list(
    timewise_dea = timewise_dea,
    timewise_dea_mt_adjusted = timewise_dea_mt_adjusted,
    training_dea = training_dea,
    training_dea_mt_adjusted = training_dea_mt_adjusted
  )
}

save(dea_results,
     file=paste0(output_dir,"mito_paper_dea_results.RData"))

# Sanity check, compare our code to pre-computed DEA
for(tissue_name in names(dea_results)){
  # load data 
  da_name = paste0("TRNSCRPT_",tissue_name,"_DA")
  curr_da = get(da_name)
  computed_da = dea_results[[tissue_name]]$timewise_dea
  computed_da = as.data.frame(computed_da)
  rownames(computed_da) = paste(
    computed_da$feature_ID,computed_da$sex,computed_da$comparison_group,sep=";"
  )
  rownames(curr_da) = paste(
    curr_da$feature_ID,curr_da$sex,curr_da$comparison_group,sep=";"
  )
  rowname_jacc = length(intersect(rownames(curr_da),rownames(computed_da))) / 
    length(union(rownames(curr_da),rownames(computed_da)))
  if(rowname_jacc < 1){
    print(paste("In tissue",tissue_name,"covered timeise data differs"))
  }
  curr_da = curr_da[rownames(computed_da),]
  inds = !is.na(curr_da$p_value) & !is.na(computed_da$p_value)
  rho1 = round(cor(curr_da$p_value[inds],computed_da$p_value[inds]),digits = 4)
  rho2 = round(cor(curr_da$logFC[inds],computed_da$logFC[inds]),digits = 4)
  print(paste(tissue_name,rho1,rho2))
}

# Merge tables and use IHW
dea_combined = list()
for(j in 1:length(dea_results)){
  print(j)
  tissue = names(dea_results)[j]
  for(nn in names(dea_results[[j]])){
    d = dea_results[[j]][[nn]]
    d = as.data.frame(d)
    d$tissue_abbr = tissue
    d$assay = "TRNSCRPT"
    if(!is.null(dea_combined[[nn]])){
      missing_cols = setdiff(colnames(dea_combined[[nn]]),colnames(d))
      for(cc in missing_cols){d[[cc]]=NA}
      d = d[,colnames(dea_combined[[nn]])]
    }
    dea_combined[[nn]] = rbind(dea_combined[[nn]],d)
  }
}
sapply(dea_combined,dim)
dea_combined$training_dea$ihw_q = IHW::ihw(
  dea_combined$training_dea$p_value,
  factor(dea_combined$training_dea$tissue_abbr),
  alph = 0.05
)@df$adj_pvalue
dea_combined$training_dea_mt_adjusted$ihw_q = IHW::ihw(
  dea_combined$training_dea_mt_adjusted$p_value,
  factor(dea_combined$training_dea_mt_adjusted$tissue_abbr),
  alph = 0.05
)@df$adj_pvalue
table(dea_combined$training_dea$ihw_q < 0.05,
      dea_combined$training_dea_mt_adjusted$ihw_q < 0.05)
sapply(dea_combined,dim)

save(dea_combined,
     file=paste0(output_dir,"ihw_adjusted_mito_paper_dea_results.RData"))

```

```{r}
load(paste0(output_dir,"ihw_adjusted_mito_paper_dea_results.RData"))
```


<!-- # GSEA before (total effects) and after adjustment -->

<!-- ```{r} -->
<!-- library(fgsea) -->
<!-- mito_gsea_res = c() -->
<!-- for(analysis_name in names(gsea_input)){ -->
<!--   print(paste("analyzing",analysis_name)) -->
<!--   m = gsea_input[[analysis_name]] -->
<!--   print(paste(analysis_name,cor(m)[1,2])) -->
<!--   gsea_adj = fgsea(mitocarta_pathways,m[,1],minSize=5,sampleSize=1000) -->
<!--   gsea_adj = as.data.frame(gsea_adj) -->
<!--   gsea_unadj = fgsea(mitocarta_pathways,m[,2],minSize=5,sampleSize=1000) -->
<!--   gsea_unadj = as.data.frame(gsea_unadj) -->
<!--   rownames(gsea_adj) = gsea_adj$pathway -->
<!--   rownames(gsea_unadj) = gsea_unadj$pathway -->
<!--   gsea_unadj = gsea_unadj[rownames(gsea_adj),] -->
<!--   colnames(gsea_unadj) = paste0(colnames(gsea_unadj),"_unadj") -->
<!--   colnames(gsea_adj) = paste0(colnames(gsea_adj),"_adj") -->
<!--   gsea_res = cbind(gsea_adj,gsea_unadj) -->
<!--   arr = strsplit(analysis_name,split=",")[[1]] -->
<!--   gsea_res$query = analysis_name -->
<!--   gsea_res$tissue = arr[1] -->
<!--   gsea_res$sex = arr[2] -->
<!--   gsea_res$week = arr[3] -->
<!--   mito_gsea_res = rbind(mito_gsea_res,gsea_res) -->
<!-- } -->
<!-- mito_gsea_res = mito_gsea_res[!is.na(mito_gsea_res$pval_unadj),] -->
<!-- mito_gsea_res = mito_gsea_res[!is.na(mito_gsea_res$pval_adj),] -->
<!-- ``` -->

<!-- ```{r} -->
<!-- qthr=0.1 -->
<!-- gsea_simple_stats = c() -->
<!-- for(query in unique(mito_gsea_res$query)){ -->
<!--   res = mito_gsea_res[mito_gsea_res$query==query,] -->
<!--   adj_fdr = res$padj_adj < qthr -->
<!--   unadj_fdr = res$padj_unadj<qthr -->
<!--   v = data.frame( -->
<!--     query = query, -->
<!--     overlap = sum(adj_fdr & unadj_fdr), -->
<!--     unadjusted_only = sum(!adj_fdr & unadj_fdr), -->
<!--     adjusted_only = sum(adj_fdr &!unadj_fdr) -->
<!--   ) -->
<!--   gsea_simple_stats = rbind(gsea_simple_stats,v) -->
<!-- } -->
<!-- gsea_simple_stats = gsea_simple_stats[order(gsea_simple_stats$query),] -->
<!-- m = as.matrix(gsea_simple_stats[,-1]) -->
<!-- rownames(m) = gsea_simple_stats[,1] -->
<!-- library(gplots) -->
<!-- heatmap.2(m,scale="none",trace="none",Colv = F,Rowv = F,col=bluered,mar=c(10,10),cexCol=1.3) -->
